src/Storages/StorageDistributed.cpp

```diff
@@ -881,6 +863,139 @@ std::optional<QueryPipeline> StorageDistributed::distributedWrite(const ASTInser
 }
 
 
+std::optional<QueryPipeline> StorageDistributed::distributedWriteFromClusterStorage(const std::shared_ptr<IStorageCluster> & src_storage_cluster, const ASTInsertQuery & query, ContextPtr local_context)
+{
+    const auto & settings = local_context->getSettingsRef();
+    auto extension = src_storage_cluster->getTaskIteratorExtension(local_context);
+
+    auto src_cluster = src_storage_cluster->getCluster(local_context);
+    auto dst_cluster = getCluster();
+
+    /// Here we will check not an equality of sourse and destination cluster
+    /// But instead we will check that the source cluster addresses are the subset of the destination cluster
```

Comment:
I think it is reasonable for select from StorageDistributed. For s3Cluster we can just take cluster from dst.
Or maybe automatically detect StorageS3 and replace it to s3Cluster.